### DISCUSSION TITLE
Namespace prefix is serialized to "ns1" despite binding it.

### DISCUSSION THREAD

**Discussion by BalduinLandolt** [RDFLib/rdflib#1609] [Q&A] [answered]
Hi, sorry to bother and to double-post (this [here](https://stackoverflow.com/questions/65818401/namespace-binding-in-rdflib)), but the more I think about it, the less I think it's intended behaviour:

In the following minimal test case:

```python
from rdflib import Graph, Namespace, Literal, RDF

base = "http://test.com/ns"
foobar = Namespace("http://test.com/ns#")
g = Graph(base=base)
g.bind('foobar', foobar)

g.add((foobar.something, RDF.type, Literal('Blah')))
g.add((foobar.something, foobar.contains, Literal('a property')))

g.add((foobar.anotherthing, RDF.type, Literal('Blubb')))
g.add((foobar.anotherthing, foobar.contains, Literal('another property')))

print(g.serialize(format='turtle').decode("utf-8"))
```

I get

```turtle
@base <http://test.com/ns> .
@prefix foobar: <http://test.com/ns#> .

<#anotherthing> a "Blubb" ;
    ns1:contains "another property" .

ns1:something a "Blah" ;
    ns1:contains "a property" .
```

what I'd expecte is more like

```turtle
@base <http://test.com/ns> .
@prefix foobar: <http://test.com/ns#> .

<#anotherthing> a "Blubb" ;
    foobar:contains "another property" .

<#something> a "Blah" ;
    foobar:contains "a property" .
```

So either there is something I fundamentally don't understand about RDFLib and how to use namespaces, or there's something funky going on.  
`ns1`, not being defined in the beginning, can't be resolved correctly, right?  
Also, I'm not clear, why `<#anotherthing>` refers to `@base` while `ns1:something` (in opposition to `<#something>`) refers to the `@prefix foobar`.

If you could tell me what I'm doing wrong, I'd be very grateful! But also, this might be an indication that the documentation on that part is not 100% clear. :)

**Comment by BalduinLandolt** [comment]
I think I found the issue: `@base` and `@prefix` point to the same URI, which it doesn't seem to like.

If I remove the base, it turns out correct. If I don't bind the namespace, it adds it as `ns1`, so the result is correct (though ugly). Just if I have both, it doesn't work.

Is this behaviour intended? If so, it should be documented.

**Comment by nicholascar** [comment, marked as answer]
As per [the answer to the same question in Stack Overflow](https://stackoverflow.com/questions/65818401/namespace-binding-in-rdflib): you need to complete the URI in the base declaration by adding the final `#`, or just don't use a base at all. As it is, the `base` is indeed clashing with the declared prefix.

There probably could be better handing of this in RDFlib but it's an extreme edge case predicated on incorrect declarations of a base that match a prefix, so I'm tempted to close the issue as I don't think anyone's likely to address it.

**Comment by BalduinLandolt** [comment]
Thank you for your reply, the final `#` indeed fixes it.  
However, this brought another small "wonkiness" to light: now I get

```turtle
@base <something#> .
@prefix foobar: <something#> .

<ting> a <thing-class> ;
    <contains> "Literal Property" .
```

So the `@prefix` is declared but never actually used (which makes sense, as it's the same as the `@base`).  
But as it is never used, I figured I don't need to bind it in the first place.  
But when I removed the binding, I got `@prefix ns1: <something#>` which again is never actually used.

In that sense, there is certainly some room for improvement... But I full well understand if that's not something that's getting addressed. It is an edge case after all.
